### PR TITLE
Catch getToken error and set token to null

### PR DESCRIPTION
**Pull request by kengreeff**
Hi team,

When updating our app to 0.46.0 our graphql calls with `@skipAuth` would no longer work as expected and would instead return `Error: Login Required`. I believe this comes from the change here:

https://github.com/redwoodjs/redwood/pull/4320

From closer inspection it looks like the Auth0 library throws an error when `getAccessTokenSilently` fails.

This PR attempts to fix this by wrapping the call in a try/catch. 

Side Note:
I managed to make the changes and test using `RWFW_PATH=~/code/redwood yarn rwfw project:sync` but was having a lot of trouble as the node_modules folder in web directory would remain stale - I could see that the node_modules folder in the root directory was updated using the `copyFrameworkFilesToProject` function from `tasks/framework-tools/lib/project.mjs` - not sure if there is a second PR there to update the node_modules in there or remove the directory entirely (which seemed to work).

### CODE REVIEW

**Comment by jtoar** [comment]
Thanks @kengreeff! We've had some reports about multiple `node_modules` directories cropping up lately. Ideally there should only be one at the root of your project. Removing it and reinstalling is the way to go if it's affecting you. In v0.48.0, we added preview support for yarn 3 preview, so we're hoping that'll help us keep things a little more consistent.

This may also be helpful: there's a flag you can pass to `yarn rw dev` to watch for changes in `node_modules`:

```
yarn rw dev   --watchNodeModules
```

**Comment by kengreeff** [comment]
Do you guys need me to do anything more with this one or are we all good?

**Comment by Tobbe** [comment]
@kengreef I'll have a look at it, hopefully tomorrow, and then I'll get back to you if there's anything I need. For now you can sit back and enjoy the feeling of having submitted an awesome PR 😎

**Comment by kengreeff** [comment]
@Tobbe lol thanks mate!

**Comment by Tobbe** [comment]
So for me the source of this issue is this line
https://github.com/redwoodjs/redwood/blob/4261bbe719c93af7025897fd5985e2f5afc618e4/packages/auth/src/AuthProvider.tsx#L27

When I made the changes in #4320 I believed that the contract for the auth providers was that `getToken` should return a promise that resolves to `null` if they can't refresh the token. If we can trust that contract my change in our Apollo code is safe. But apparently we can't, as made evident by this PR and also https://github.com/redwoodjs/redwood/pull/4539

We need to go through all places we use `getToken` (apollo/index.ts and AuthProvider.tsx from what I can see) and figure out what kind of behavior we want. And then we should document that. A comment here https://github.com/redwoodjs/redwood/blob/4261bbe719c93af7025897fd5985e2f5afc618e4/packages/auth/src/AuthProvider.tsx#L27 would be good for that I think.

And then we need to go through all auth providers to make sure they follow that behavior. If we decide they should *not* throw, maybe we can add a try/catch for that in https://github.com/redwoodjs/redwood/blob/4261bbe719c93af7025897fd5985e2f5afc618e4/packages/auth/src/AuthProvider.tsx#L208-L210 that way we don't have to add it to every single provider implementation. (And then we should probably also remove it from those that have it to reduce redundancy.) If we decide they *should* throw, we need to go through them all to add that.

dbAuth seems to return `''` if it can't fetch a token. That should probably change to `null` to better align with the other providers.

We need tests for this stuff. This is the second time this happens now 🙁  (the first time being #4539). 

@kengreeff Do you have time to help me sort all of this out? Sorry for the can of worms 🙈

**Comment by kengreeff** [comment]
Haha how good is pulling on the string and seeing what unravels! 

I'm keen, let me know the next steps

**Comment by kengreeff** [comment]
Hey @Tobbe what would you like me to get started on to keep this one moving forward?

**Comment by Tobbe** [comment]
@kengreeff Sorry dude, I got swamped with day job stuff (that extended into evening/night job 🙁 )

So, actions:

1. Go through all uses of `getToken` to try to understand what we want the contract to be (i.e. possible return values and `throw` behavior)
2. Document the wanted behavior as a code comment in `AuthProvider.tsx` https://github.com/redwoodjs/redwood/blob/4261bbe719c93af7025897fd5985e2f5afc618e4/packages/auth/src/AuthProvider.tsx#L27
3. Update all auth providers to follow what we documented in step 2.
4. Figure out how to add tests for this stuff, to catch any future regressions.

Is that clear? Let me know if you want me to elaborate on anything.

**Comment by kengreeff** [comment]
Great, thanks - I will try take a look this weekend

**Comment by kengreeff** [comment]
# getToken implementations in auth providers

## Auth0

```js
Promise<string | GetTokenSilentlyVerboseResponse>
```

When using `getTokenSilently` Auth0 can throw an error if login is required. I think we can safely catch this and ignore in the authClient. Below is from their docs. It looks like they only throw if the error is not `login_required`

```js
try {
  await auth0.getTokenSilently();
} catch (error) {
  if (error.error !== 'login_required') {
    throw error;
  }
}
```

## Azure Active Directoy

```js
Promise<string | null> (or redirects)
```

Currently trying `acquireTokenSilent` and catching if there is an error and using `acquireTokenRedirect`. Code could probably be updated to follow official docs.

```
try {
  const token = await client.acquireTokenSilent(request)
  return token.idToken
} catch (err) {
  if (err instanceof InteractionRequiredAuthError) {
    // fallback to interaction when silent call fails
    return client.acquireTokenRedirect(request)
  }
}
```

## Clerk

```js
Promise<string | null>
```

Returns the clerk client session ID or null

## DB Auth

```js
Promise<string | null>
```

Looks like dbAuth returns a string (token) or null

```js
if (token.length === 0) {
  return null
} else {
  return token
}
```

## Ethereum

```js
Promise<string | null>
```

Uses `localStorage.getItem()` which will return null or the token.

## Firebase

```js
Promise<string | null>
```

Hard to find the method in the source code that actually retrieves the token but it looks like it is usually a string.

## Go True

```js
Promise<string | null>
```

Currently fetching JWT and catching errors and returning null.

## Magic Link

```js
Promise<string | null>
```

It looks like this will actually return `undefined` as well. I am new to typescript but unless the types set the value this is undefined:

```js
let token: string | null
```

It looks like this method `client.user.getIdToken()` can return an error if it cant get a token:

https://github.com/magiclabs/magic-js/blob/af0a37d3f457b4d61fa84a17bb278d078159814d/packages/%40magic-sdk/provider/src/modules/base-module.ts#L22

So it might be a good idea to wrap this in a try/catch as well.

## Netlify

```js
Promise<string | null>
```

Request to get token is wrapped in a try/catch. Returns null or token

## NHost

```js
Promise<string | null>
```

Returns a string or null

```js
public getJWTToken(): string | null {
  return this.currentSession.getSession()?.jwt_token || null;
}
```

## Supabase

```js
Promise<string | null>
```

Returns access_token or null

## Super Tokens

```js
Promise<string | null>
```

Looks like this method can throw, maybe wrap in a try/catch.

```js
static async getAccessTokenPayloadSecurely(): Promise<any> {
  return Session.getInstanceOrThrow().getAccessTokenPayloadSecurely();
}
```

# AuthProvider

As you mentioned, it is probably safest to update as below so that any unexpected errors are caught here (as libraries updated etc):

```js
getToken = async () => {
  let token

  try {
    token = await this.rwClient.getToken()
  } catch (error) {
    token = null
  }

  return token
}
```

File: packages/auth/src/AuthProvider.tsx [hunk 1]
```diff
@@ -219,7 +219,15 @@ export class AuthProvider extends React.Component<
   }
 
   getToken = async () => {
-    return this.rwClient.getToken()
+    let token
+
+    try {
+      token = await this.rwClient.getToken()
+    } catch (error) {
```
Comment (Tobbe) [inline packages/auth/src/AuthProvider.tsx:226]:
```suggestion
    } catch {
```

Since we're not using the error there's no need to declare it

**Comment by Tobbe** [comment]
> I have gone through all the auth providers and documented above.

Thanks @kengreeff, I really like how thorough that was!

With your change to the code we can get rid of some of the try/catches inside some of the clients, right? Like Netlify for example.

> Havent had a look at a test just yet - what were you thinking we could test?

You stated that the error you found was this:

> When updating our app to 0.46.0 our graphql calls with `@skipAuth` would no longer work as expected and would instead return `Error: Login Required`.

Could you write a test for that? A test that tries to access a resolver with `@skipAuth` on it. It should fail without your fix, and pass with your fix in place. I guess the problem is going to be to setup Auth0 (or some other provider) for the test case. Not sure, but maybe you can find some inspiration here: https://github.com/redwoodjs/redwood/pull/4673 Or maybe (which would be preferred) there is a much simpler way to test this that I haven't though of yet. Could you see if you can come up with something? Let me know if you get stuck and I'll think on it some more too, alright?

**Comment by kengreeff** [comment]
Alright @Tobbe hopefully this is it.

I have added a test to the AuthProvider component that ensures that we always get a return value from `getToken` and dont have to catch. 

I am not sure you can test `@skipAuth` directly from what I can see, but I think this covers it as it is the root of the problem.

If you swap out to the old code the test fails as it has to catch and the token is set to `FAIL` instead of `null`

**Review by Tobbe — COMMENTED** [review]
I made some code edit suggestions on the test case.
Mainly I wanted to get rid of `act` (because it's very rarely needed). And in doing so I also realized I could get rid of some other code to make the test case more focused.

**Comment by Tobbe** [comment]
With your code changes in place, what do you think about the existing code in `goTrue.ts` and `netlify.ts`?  Can we remove the try/catch in their `getToken` implementations?

File: packages/auth/src/__tests__/AuthProvider.test.tsx [hunk 2]
```diff
@@ -2,7 +2,7 @@ require('whatwg-fetch')
 
 import { useEffect, useState } from 'react'
 
-import { render, screen, fireEvent, waitFor } from '@testing-library/react'
+import { act, render, screen, fireEvent, waitFor } from '@testing-library/react'
```
[the remark this answers is not in the document]
Comment (kengreeff) [reply packages/auth/src/__tests__/AuthProvider.test.tsx:5]: eslint was asking to wrap setState calls in `act`

File: packages/auth/src/__tests__/AuthProvider.test.tsx [hunk 3]
```diff
@@ -866,3 +866,60 @@ test('proxies validateResetToken() calls to client', async () => {
...
+
+  render(
+    <AuthProvider client={mockAuthClient} type="custom">
+      <TestAuthConsumer />
+    </AuthProvider>
+  )
+
+  await waitFor(() => screen.getByText('getToken Complete'))
+
+  expect(screen.getByText('Token: null')).toBeInTheDocument()
```
[the remark this answers is not in the document]
Comment (kengreeff) [reply packages/auth/src/__tests__/AuthProvider.test.tsx:924]: just want to confirm you ran the code with your changes?

**Comment by kengreeff** [comment]
> With your code changes in place, what do you think about the existing code in `goTrue.ts` and `netlify.ts`? Can we remove the try/catch in their `getToken` implementations?

I think that the clients should catch their own errors and return null - the new change should be a final safety, in my opinion. It is expected that they return null or token, not throw

**Comment by Tobbe** [comment]
> I think that the clients should catch their own errors and return null - the new change should be a final safety, in my opinion. It is expected that they return null or token, not throw

Great! 🙂 

Could you please add a comment to both AuthProvider.tsx L26 and to the new code in `getToken` saying this, so we all remember the decision when looking back at this code?

**Comment by kengreeff** [comment]
busy making your changes now

**Comment by kengreeff** [comment]
> ❌ Deploy Preview for _redwoodjs-docs_ failed.
> 
> 🔨 Explore the source changes: [6a30c5e](https://github.com/redwoodjs/redwood/commit/6a30c5e10ed8efb7b8ad3a1607e95af66482986c)
> 
> 🔍 Inspect the deploy log: https://app.netlify.com/sites/redwoodjs-docs/deploys/6232e317b78e550008142e43

Why would this be failing?

**Comment by Tobbe** [comment]
> Why would this be failing?

We're in the process of moving our docs over to this repo. I'll try merging the latest changes from `main` into this branch and see if that solves it.

**Comment by jtoar** [comment]
@kengreeff If you update from main it should build ok. I just merged a PR in that adds all the docs to the main repo (https://github.com/redwoodjs/redwood/pull/4749); since I enabled deploy previews, netlify tries deploying every PR now, but this PR doesn't have the docs at all so it fails.

**Comment by kengreeff** [comment]
Thanks guys :) @Tobbe @jtoar

**Comment by Tobbe** [comment]
Thanks for your work on this @kengreeff! Really appreciate it 🙏

**Pull request merged**